cardano-node/src/Cardano/Tracing/Metrics.hs

```diff
@@ -68,9 +78,31 @@ instance HasKESMetricsData (ShelleyBlock c) where
 
 instance HasKESMetricsData ByronBlock where
 
-instance HasKESMetricsData (SimpleBlock a b) where
-
-instance (HasKESMetricsData x, NoHardForks x)
-      => HasKESMetricsData (HardForkBlock '[x]) where
-  getKESMetricsData forgeState =
-    getKESMetricsData (project forgeState)
+instance All HasKESMetricsData xs => HasKESMetricsData (HardForkBlock xs) where
+  getKESMetricsData _ forgeStateInfo =
+        combineKESMetrics
+      . hcollapse
+      . hcmap (Proxy @ HasKESMetricsData) getOne
+      . OptNP.toNP
+      . getPerEraForgeStateInfo
+      $ forgeStateInfo
+    where
+      getOne :: forall blk. HasKESMetricsData blk
+             => (Maybe :.: WrapForgeStateInfo) blk
+             -> K KESMetricsData blk
+      getOne (Comp mbWrapForgeStateInfo) = K $
+          case unwrapForgeStateInfo <$> mbWrapForgeStateInfo of
+            Nothing -> NoKESMetricsData
+            Just oneForgeStateInfo ->
+              getKESMetricsData (Proxy @ blk) oneForgeStateInfo
+
+      -- Multiple eras can have 'KESMetricsData', e.g., Shelley and Shelley +
+      -- native assets. We don't know which era we're in, so we have to make a
+      -- choice: we choose the last era's 'KESMetricsData'.
+      --
+      -- TODO this might need revision when we have updated everything for the
+      -- first hard fork after Shelley, but before it actually happened. At
+      -- that moment, we'll trace the KESMetricsData of the upcoming era,
+      -- instead of the KESMetricsData of the current Shelley era.
+      combineKESMetrics :: [KESMetricsData] -> KESMetricsData
+      combineKESMetrics = lastDef NoKESMetricsData
```

Comment:
This will need some attention when we do the next hard fork.

Reply:
After thinking a little about this: when we do the next hard fork, we might as well use the same KES key for the current and the next era, in which case this doesn't matter much.